open-craft/edx-platform
# Store Collection metadata + component count in meilisearch

## Description

This PR makes several enhancements related to how Collections are indexed in the Studio Search app.

1. Handles the `LIBRARY_COLLECTION_CREATED` / `LIBRARY_COLLECTION_UPDATED` events and adds/updates the collection document in the Studio content search index.
2. Stores the Collection.key in the document `block_id` field. This key + the library_key is required for accessing the Collection REST APIs.
3. Adds a new field to the document called `num_children`. Collections store the number of entities in the collection.
   Future PRs may store XBlock children counts here.

Libraries v2 is not used in production by default yet, so this change shouldn't affect any current users.

## Supporting information

Related to https://github.com/openedx/frontend-app-course-authoring/issues/1260
Based on https://github.com/openedx/edx-platform/pull/35469
Private-ref: [FAL-3787](https://tasks.opencraft.com/browse/FAL-3787)

## Testing instructions

Set up tutor:

* Setup tutor with this PR, https://github.com/openedx/frontend-app-course-authoring/pull/1259, and the https://github.com/open-craft/tutor-contrib-meilisearch plugin
* Ensure you are running with the latest versions of openedx-learning + openedx-events.
* Run migrations: `tutor dev run cms python manage.py migrate`
* Run `tutor dev run cms ./manage.py cms reindex_studio --experimental` to add the new fields.

In the content-authoring MFE:

* Create a library, e.g. `lib:SampleTaxonomyOrg1:AL1`
* Add a collection to that library, e.g title="My Collection"
* Add blocks to your library, and note their usage keys.
  Hint: you can retrieve the block keys using the REST API, e.g. http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/blocks/

In the shell `tutor dev run cms python manage.py cms shell`, add blocks to the collection:
```python
from opaque_keys.edx.keys import UsageKey
from opaque_keys.edx.locator import LibraryLocatorV2
from openedx.core.djangoapps.content_libraries import api                                                                                         
lib_key_str = "lib:SampleTaxonomyOrg1:AL1"
collection_key = "my-collection"
block_key_strs = [
    "lb:<your block key",
    "lb:<your block key",
    # ....
]

library_key = LibraryLocatorV2.from_string(lib_key_str)
block_keys = [
    UsageKey.from_string(key_str) for key_str in block_key_strs
]

api.update_library_collection_components(library_key, collection_key, usage_keys=block_keys)
```

In meilisearch http://meilisearch.local.edly.io:7700:

* Login using the api key: `tutor config printvalue MEILISEARCH_API_KEY`
* Search for your collection title and ensure you find it in the search index.
* Check the following fields:
  `id = collection.id` (unique)
  `block_id = collection.key`
  `num_children = <number of blocks in the collection>`

## Deadline

Before Sumac gets cut in October.
